### ISSUE TITLE
Defining "1.0.0"

### ISSUE THREAD

**Issue by jgeewax** [googleapis/google-cloud-common#49]
# Defining what `1.0.0` means
## TL;DR

`1.0.0+` means:
1. Coverage of Datastore, Storage, Pub/Sub, BigQuery, and Logging.
2. Fast (and measured) response time on GH issues filed
3. Example snippets for all methods
4. Docs published as structured JSON document
5. Support for *nix, Mac, and Windows

Assuming you agree with that, here's where (I think) we are:

| Library | Core | Support | Docs | OSs |
| --- | --- | --- | --- | --- |
| `gcloud-golang` | 20+% | ? | ? | ✓ |
| `gcloud-java` | 60% | ? | 0% | ✓ |
| `gcloud-node` | ✓ | ✓ | ✓ | ✓ |
| `gcloud-php` | 20% | ? | ✓ | ? |
| `gcloud-python` | ✓ | ✓ | 10% | 75% |
| `gcloud-ruby` | ✓ | ✓ | 100% | 75% |
### To-do list (for Mar 31, 2016):
- Core:
  - [x] Finish Logging in gcloud-ruby
  - [x] Finish Logging in gcloud-python
- Support:
  - [ ] Instrument some way of measuring our response times on tickets
- Docs:
  - [x] Define the JSON schema for docs
  - [x] Finish the gcloud-common shared Angular app and deployment scripts
  - [x] Export gcloud-node docs into new shared schema
  - [ ] Export gcloud-python docs into new shared schema
  - [ ] Export gcloud-java docs into new shared schema
  - [x] Export gcloud-ruby docs into new shared schema
- OSs
  - [x] Make sure gcloud-node works on Windows
  - [ ] Make sure gcloud-python works on Windows
  - [ ] Make sure gcloud-java works on Windows
  - [ ] Make sure gcloud-ruby works on Windows
## Goal

Getting to a `1.0.0` version has been on the list of stuff to do for quite some time, and we've discussed a bit about what merits that version number, but never put it down in writing. The goal here is just to finalize what the criteria are for labeling something as "`1.0.0`".
### Compatibility

According to semver.org, `1.x.y` should be compatible (ie, require no code changes) for all `x` and `y`. This means if we jump to `1.0.0` on a library, and we then need to modify the API surface in a way that breaks existing code, we need to bump the version to `2.0.0` for that change.
### Core coverage

Part of `1.0.0` is technical, but another part is about "what can be done" with the library. Based on some internal usage statistics, I'm proposing that the following Google Cloud APIs should be "done" in order to increment our major version (note that all of these APIs are application-focused rather than infrastructure-management focused):
1. **Google Cloud Datastore**: Many people using Google Cloud started with App Engine and therefore have data in App Engine. We need a way for those customers to interact with their data in App Engine from outside App Engine (ie, from Compute Engine)
2. **Google Cloud Storage**: A key use case for new developers is flat file storage for user uploads, and Google Cloud Storage solves this problem. Although GCS and S3 are XML-API compatible (ie, you can use boto in Python to talk to GCS), I think it's a good idea to provide a consistent story for this common situation rather than asking users to install and use two libraries, two snippets of authentication code, etc.
3. **Google Cloud Pub/Sub**: Although not as popular as some of the other APIs, Pub/Sub is becoming increasingly important for app developers, particularly those exploring Cloud DataFlow and using Pub/Sub as a point for data ingestion.
4. **BigQuery**: While the querying API for BigQuery is less "application-focused", for customers looking to provide pretty summary graphs to customers, this is quite important. Further, the data ingestion API is seeing tons of traffic, so this seems like a critical (and relatively straight-forward) service to require.
5. **Google Cloud Logging**: While not as important for the piecemeal users (ie, "I just want BigQuery"), many people trying to use the full integrated stack on Google Cloud want an integrated and simple logging solution (sending to `/var/log` and shipping those files to GCS is not exactly "fun"). To support those users we really need good integrated logging clients.
### Support and responsiveness

Whatever we ship is really only as good as how well we support it. If it takes us 4 weeks to acknowledge a bug in the library or answer a question about usage, we're basically saying "here... you deal with it".

This should be a conversation, but I think we should shoot for:
1. Acknowledgement
   - 50% of issues are ack'd within 3 business days
   - 100% of issues are ack'd within 5 business days
2. Resolution
   - Resolving issues can significantly vary in complexity and difficulty, so putting an ETA on bug fixes as a goal can get messy.
   - Putting priorities or a "critical" label on an issue is a very subjective thing, so that can also get messy.
   - If anything, it's important that we make _progress_ on resolving issues. In that sense, it's more important that we resolve a certain number of issues in a given time frame (assuming there are at least that many issues) rather than putting an ETA for any given issue.
   - Based on that, I'd propose that we aim to get at least 5 user-reported issues resolved per week (assuming there are more than 5 outstanding issues)
   - Code review is real work too, so reviewing and merging external pull-requests should count as "resolving an issue"
### Documentation

Our library is really only as good as the documentation around it, so a few goals focused on a great experience with documentation is important.
1. All docs should be exported as JSON to be consumed by one common application.
   - A side effect of this is that external sources should be able to consume our JSON documentation and get data for things like "I have a Google API method called `storage.objects.delete`. Show me the example in gcloud-node for that method."
   - See https://github.com/GoogleCloudPlatform/gcloud-common/issues/33 for more info
2. All our API methods should have usage examples right next to them.
3. Any example code should be tested such that if an example doesn't work, a build must fail.
### Cross-platform support

As (primarily) *nix developers, we tend to forget about other platforms. To jump to `1.0.0`, our libraries should be tested and have a great (simple and error-free) installation and usage story for more than just Linux. I'm proposing we have the same quality of experience for the following OS's:
1. Mac OS X
2. Ubuntu
3. RHEL / CentOS
4. Windows

(All of these should be relatively easy for us to do with the exception of Windows, the it could be made slightly easier using AppVeyor.)
## Outstanding questions
- If we call gcloud-node 1.0, and how do we add a new API surface that we wouldn't necessarily call 1.0?
  - 1.1-preview?
  - separate artifact? (ie, `gcloud-newapi @ 0.1` ?)

---

/cc @stephenplusplus @callmehiphop @dhermes @tseaver @blowmage @quartzmo @dwsupplee @aozarov @mziccard @ajkannan @okdave

**Comment by dhermes** [comment]
Luckily AppVeyor makes testing on Windows fairly simple

**Comment by blowmage** [comment]
I'm :+1: for the criteria defining 1.0.0. (As is @quartzmo, but he's out the rest of the week, so I'll speak for him.) For the ruby libraries, we think eventually moving to individual packages for each service would be ideal, as it would allow for us to introduce new services without changing the released API.

I've spent some time thinking about individual packages for gcloud services and how to version them. On one hand, I think it would be interesting to match the package version numbers to the cloud services, but there are many downsides to that. I think keeping all released packages in sync is the best. So a gcloud-ruby 1.1.0 release would also contain gcloud-datastore 1.1.0 and gcloud-storage 1.1.0, etc...

However, the individual packages approach does greatly complicate the angular app for showing documentation. Lets say we issue a patch release of gcloud-storage and bump that package's version to 1.0.1. We don't need to bump the main gcloud package, or all other service packages, just the storage one. How does one choose the patch version when viewing the storage documentation?

**Comment by callmehiphop** [comment]
@blowmage that's a good question. 

I think that the most common use case will just be switching versions for all the services (as it is now), so IMO we want to keep that functionality. We *could* add a secondary switcher to the individual service pages, but I could see that being confusing and leading to a bad user experience.

Maybe we should create an additional menu that allows the user to switch between the main module and the individual modules. Then when we're in an individual module state, instead of showing APIs, we show only the pages for the active module..?

Where exactly do we stand on moving to individual packages? It's been brought up a couple of times for Node and always ended up getting shot down.

**Comment by dhermes** [comment]
:+1: To secondary switcher. De-coupling services --> de-coupling releases --> decoupled versions

**Comment by blowmage** [comment]
Right now our 2016 goals are adding ~10 new services to gcloud-ruby. Our current approach is to increment the minor version each time we do, but I can see that being more difficult once we hit 1.0.0. We haven't introduced many API breaking changes pre 1.0.0, but if we needed to then we might be at version 7.0.0 by the end of the year. Having some services under active development with pre-1.0.0 individual releases may help with dealing with those changes.

**Comment by ajkannan** [comment]
:+1: for the 1.0.0 criteria and decoupling services/releases/versions

**Comment by mderka** [comment]
The 1.0.0 definition looks good.

I like the idea of secondary switcher for doc.

**Comment by csells** [comment]
I'm a fan of the 1.0.0 definition except that I believe that Windows needs to be a 1st class supported citizen along with Linux/Mac.

**Comment by jgeewax** [comment]
@csells : Can you point out where it says Windows wouldn't be a 1st class citizen? I know it's last on the list, but it's there :)

**Comment by csells** [comment]
tl;dr: I'd like using the libs from Windows to be as simple as using them from any other OS.

Under Cross-platform support, it says that every OS should be "simple" except Windows: "Windows (may not be simple, possibly easier with AppVeyor)". I'd like to see it say "Windows (should be simple)".

Or, it may be that I'm conflating the UX of libs from Windows (which should be just as simple as everything else) with the difficulty of testing the libs from Windows (which can be hard -- that's what they pay us for : )

**Comment by jgeewax** [comment]
Sorry, in parens there was meant to be "the work we have to do" should be simple / not simple -- not saying anything about what the user should experience. I'll update to make that more clear.

**Comment by stephenplusplus** [comment]
> If we call gcloud-node 1.0, and how do we add a new API surface that we wouldn't necessarily call 1.0?

I think this is giving too much weight to the term 1.0. Strictly semver speaking, 1.0 means every increment until 2.0 won't force a user to change their code. We will absolutely follow those rules, but all users should have access to new features, without being forced to install gcloud from a more obscure location (whether a separate module or a version with a `-pre` suffix).

The most straightforward path is when releasing a new API, we bump the minor. Anyone depending on >1 & <2 (npm: `^1.0.0`) will automatically pick up the new API; it's there if they want it, but if not, their existing code will continue functioning as usual. If we release a feature that turns out to have flaws, we just fix and increment the patch, so everyone gets it.

--

To me, modularizing is still a win for a lot of reasons, but it's kind of an anti-win when we have to maintain a wrapper package that versions along with the modules. So, gcloud-storage introduces a breaking change. That means the parent package `gcloud` has to have its major incremented as well, which is just a lot of bookkeeping and will inevitably lead to mismatches. Bug reports would be coming in from multiple repos, and there will be confusion over which version of which package a user is running.

**Comment by jgeewax** [comment]
> all users should have access to new features, without being forced to install gcloud from a more obscure location

The concern for me is something like the following happening:

- Google launches Google Cloud Whatever (GCW), versioned as `v1alpha1`.
- We're at version `1.0.4`, happily chugging along.
- We add support for GCW `v1alpha1` in `v1.0.5`.
- Google changes GCW significantly but makes it generally available as GCW `v1`

If we cannot make our code backwards compatible, what version should we use to launch support for GCW `v1`?

- `1.0.6` would break the assumption that all code written against `1.0.5` would still work, and it's not a patch so this doesn't make sense.
- `1.1.0` would still break the above assumption but at least follow closer to semver.
- `2.0.0` seems like a needlessly large jump (just because GCW was added), and might scare people into thinking more than just GCW support is broken moving forward... (even though all their existing code will work, but there's new stuff...)

The way I figured would make sense is...

Until something is `1+.x` (ie, `1.0.3`) worthy (we commit to backwards compatibility), it should live in a `0.x` release (ie, `0.2.14` or a `1+.x-pre` release (ie, `1.4.1-pre`).

This means the above scenario would be (remember, we were at `1.0.4` in the example):

1. Google launches GCW `v1alpha1`
  - We deploy support for it as `1.1.0-pre`
1. Google launches GCW `v1` (GA)
  - We deploy support for it as `1.1.1-pre`
1. We decide we're ready to be backwards compatible:
  - We deploy support as `1.1.1` (just drop the `-pre`)

This also means that the only reason to jump to `2.0` is if one of our existing `1.0` libraries does something backwards incompatible, for example:

- We're at `1.0.4`
- Google launches Cloud Datastore `v2alpha1` which requires us to change tons of code
  - We launch support as `2.0.0-pre`
- Google makes GCD `v2` generally available with more changes
  - We launch support as `2.0.1-pre`
- We want to support GCD `v2` and promise it'll be backwards compatible:
  - We launch support as `2.0.1` (dropping the `-pre`)

Thoughts?

**Comment by stephenplusplus** [comment]
- Google launches Google Cloud Whatever (GCW), versioned as v1alpha1.
- We're at version 1.0.4, happily chugging along.
- We add support for GCW v1alpha1 in ~~v1.0.5~~ **v1.1.0 (this is a feature, not a fix)**
- Google changes GCW significantly but makes it generally available as GCW v1

> If we cannot make our code backwards compatible

Would the API endpoint for v1alpha1 disappear? If it does, then our notice "You are using an alpha version of Google Cloud Whatever. There is no guarantee..." will have to suffice. They are using an alpha version after all, so if it disappears, there's nothing we can do about it, meaning backwards compatibility is out the window.

> what version should we use to launch support for GCW v1?

If it doesn't break the code and we can make the transition seamless, we bump the minor, putting us at 1.2.0.

If it removed features or would break the code we already offer for those connecting with GCW v1alpha1, we have to bump the major, putting us at 2.0.

> 2.0.0 seems like a needlessly large jump (just because GCW was added), and might scare people into thinking more than just GCW support is broken moving forward... (even though all their existing code will work, but there's new stuff...)

In Node, there are some tools that tell you "Hey, an update's available for your gcloud dependency." You can just blindly update, but you'll probably make the effort to read our release notes before upgrading. That'll clear up what's different.

It's a common pattern at least in Node world. Packages shift often, especially with all of the changes that came with ES6 support. Some libraries are using new language features (Node v4.0+) and releasing it under a major version upgrade, so if you're using a newer Node version, that supports the code they're using you can upgrade. If not, you stay with the older version.

> We decide we're ready to be backwards compatible:

I don't understand the difference. Every release is intended to be backwards compatible following major.minor.patch, no decision required.

Given the scope of our library and all of the services we support, I don't think it's a fair goal to try to stay modest with our versioning increments. Even without changes or additions to the services we support, our libraries are going to be shifting in breakable ways, e.g. when we switch to promises, our library will have to bump the major (maybe to 2 or 3.0 by then).

My overall point is I just think trying to mix code logistics with the presentation value of a version number makes things unnecessarily complicated. `breaking.feature.fix` describes all of the changes we should need to make. Just my opinion on the matter, though. Anyone else?

**Comment by jgeewax** [comment]
There's a nuance here that I think we may have skipped over.

The crux for me is that it seems wasteful to drag everyone along to a new major version (where they lose the compatibility guarantee across all services) just because we wanted to add in something volatile, but acted as though it wasn't (GCW `v1alpha`).

In other words, I'm suggesting that if we want to add something volatile (ie, GCW `v1alpha1`) we should make it available as a separate package (`gcloud-whatever@0.1.0`). If people want the latest bundle of all the "pre-release volatile stuff" , install `gcloud@1.x.y-pre`.

Once it settles and stops being volatile (... typically loses its `alpha` or `beta` suffix and we're confident about the API surface), we just pull it into the bundle package as you mentioned above (uptick the middle digit) and people get to install `gcloud@1.2.0`.

**Comment by callmehiphop** [comment]
> In other words, I'm suggesting that if we want to add something volatile (ie, GCW v1alpha1) we should make it available as a separate package (gcloud-whatever@0.1.0).

I like this!

> If people want the latest bundle of all the "pre-release volatile stuff" , install gcloud@1.x.y-pre.

I don't really get this part though. Since there will have to be some instruction on how to install the pre package, why not just tell the user to install the alpha package along side the main package? Then once we've moved away from the alpha version of the API  we integrate it into our main gcloud package as a new feature.

**Comment by dhermes** [comment]
Whatever command installs the whole kitchen sink could leave out the unstable parts:

```
$ pip install gcloud[all]
...
installing gcloud-datastore...
installing gcloud-pubsub...
installing gcloud-storage...
etc.
definitely not installing gcloud-whatever...
```

but then a user can get "new" / unstable packages via either directly asking for them (since we'd have unbundled packages/services) or by using some similar means to getting the whole sink:

```
$ pip install gcloud-whatever
$ # OR
$ pip install gcloud[pre]
...
installing gcloud-whatever...
...

**Comment by jgeewax** [comment]
> I don't really get this part though.

I could be wrong here, but...

Let's say I want to start using `gcloud-node` as I will when `1.2.0` is "official". It's not yet, and currently to use gcloud-whatever, I have to install `gcloud-whatever@0.1.0` and then do a separate require:

```js
var gcloud = require('gcloud');
var gcloudWhatever = require('gcloud-whatever');
```

But what I want is:

```js
var gcloud = require('gcloud');
var whatever = gcloud.whatever();
```

I was under the impression that there's no way to get the kitchen sink with the unstable stuff and write code like the second example... If there, great. Otherwise, one way to do it is by offering a `1.2.0-pre` package of "how things will likely look in version 1.2.0".

**Comment by jgeewax** [comment]
@dhermes : So long as `pip install gcloud[all]` == `pip install gcloud`, that LGTM.

**Comment by callmehiphop** [comment]
@jgeewax What's wrong with requiring the second module? I think if anything, explicitly requiring the `gcloud-whatever` package would imply that it's unstable. If we want to the user to be able to access the `whatever` service via the main package, then it should be in the main package. IMO trying to magically combine modules will make our library less predictable to users.

I'm not a huge fan of the pre tag, again from a user perspective, it would make using an alpha API a little more complicated to get started with. As a Node developer (we embrace Semver), trying to preserve the major version number isn't really something that I think should be prioritized, especially if it's going to make installation a bit more tedious.

**Comment by stephenplusplus** [comment]
I'm concerned about splitting out only some services while their API is becoming generally available. If published separately, at the time they reach stability, what do we do-- copy and paste the code into gcloud-node, and then what about the other module? We can deprecate it, close the issues on the repo, and put a link to gcloud-node's issues page in the readme? I think it's better to just keep it in the core repo.

With the pre-release tags, what happens in the scenario GCW alpha->beta->GA takes longer than our next regular release?

In other words, we release `1.1.0-alpha` with support for GCW alpha. But we also just added in a feature that starts your coffee maker if you're downloading a large file from GCS, and it's ready to go, so we want to release `1.1.0`. By the time GCW is stable and ready for entry into the main branch, we might be on `1.6.0`... so `1.1.0-alpha` isn't describing 1.1 at all. Do we keep releasing `1.2.0-alpha`, `1.2.2-alpha`, `1.3.0-alpha`...`1.7.0-alpha` so that someone using GCW can still keep up to date with the changes in the main stream?

I think prerelease tags are mostly for "make sure this doesn't break existing codebases. Give it a week, see if we get bug reports, then consider it safe." But that doesn't describe our use case.

It seems easier to just go with the grain with semver, and admit that it's not a great system for pre-release APIs, but it's [what we have](http://www.jongleberry.com/semver-has-failed-us.html).

**Comment by callmehiphop** [comment]
@stephenplusplus perhaps instead of copypasta, once an external package is no longer using an alpha API we could install it as a dependency with a fixed version? I'm sorta really into the idea that we would have separate packages for APIs.. since if a user only wanted `gcloud-whatever` they could get away with only installing that particular package.

**Comment by stephenplusplus** [comment]
I'm still not [sold on modularizing](https://github.com/GoogleCloudPlatform/gcloud-common/issues/49#issuecomment-174517200), but if we're going that way and can [resolve all of our issues that come along with it](https://github.com/GoogleCloudPlatform/gcloud-node/issues/91#issuecomment-138522547), we would want to make sure we modularize the rest of our library as well.

**Comment by quartzmo** [comment]
:+1: for the Core, Support, Docs, OSs criteria! Very well explained @jgeewax, thanks for the clear writeup.

Regarding modularization and versioning, I have been enthusiastic about modularization, but am starting to reconsider. As hard as I try, I just can't imagine enough downside in practice to the single library approach. The issues that have been raised above seem quite thorny (regarding documentation versioning by @blowmage and everything @stephenplusplus has written and linked). I'm going to continue to research and think about it.

**Comment by dhermes** [comment]
Positives (of unbundling) for `gcloud-python` are:

- Much smaller downloads for users that only use one service
- Separation of testing concerns (faster CI)
- Decoupled release schedules

**Comment by blowmage** [comment]
The core problem @quartzmo and I see here is that versioning for libraries such as gcloud is not a solved problem. Similar libraries either claim to use semver, but don't, or use a different versioning approach (e.g. rails).

For example, the aws ruby package is a monolithic package that includes all code to connect to all services in a single library. There are no smaller packages. It declares it [uses semver](https://github.com/aws/aws-sdk-ruby#versioning), but it does not. It is not hard to find breaking changes in [patch releases](https://github.com/aws/aws-sdk-ruby/releases/tag/v2.1.23).

The azure npm package is an "umbrella" package: It declares fixed dependencies on the individual service packages. Because the umbrella lags behind, however, installing it does not get you the latest versions of each individual package.The last umbrella release was in March, 2015, so perhaps that community has found the umbrella less useful than the individual packages.

What azure npm shows is that separate packages can be developed, versioned, and released independently. Compare the versions specified in the [latest umbrella release](https://github.com/Azure/azure-sdk-for-node/blob/v0.10.5-March2015/package.json) with the current versions of those individual service packages. We can't link to all [32 packages](https://www.npmjs.com/package/azure), but they are several releases newer than the fixed dependency.

We would prefer to move to individual packages, while keeping an umbrella package. However, the more we look at this, the less favorable we are on keeping all package versions in sync. The simplest approach that we see is to use semver for individual packages, and not use semver for the umbrella. It would be possible to follow semver on the umbrella, but that would require a major release of the umbrella for every major release of an individual package.

**Comment by blowmage** [comment]
Regarding core 1.0.0 criteria, I would absolutely love to get GRPC support added to the definition of 1.0.0. Assuming the installation of the grpc packages is "solved" and no longer require libgrpc to be installed or the package to be compiled. I think moving from REST to GRPC will require some API changes to our libraries, and I would rather get those changes in before 1.0.0.

**Comment by pcostell** [comment]
I'd like to add one more constraint to a gcloud 1.0.0 release which is a final review by the backend teams (if they choose to do so). We (the Datastore team) did initial reviews before the gcloud libraries got off the ground, I think it's a good idea to go back before a 1.0.0 cut and ensure that there aren't any bad design decisions that are getting locked in. There should be a time window (maybe 2 weeks?) in which backend teams can do a final review and recommend any necessary changes prior to 1.0.0.

**Comment by blowmage** [comment]
A final review sounds good to me. :+1:

**Comment by tbetbetbe** [comment]
:+1:  [modularization of gcloud](https://github.com/GoogleCloudPlatform/gcloud-common/issues/49#issuecomment-173008535), I think that makes a lot of sense
  
FYI googleapis folks: @anthm, @vsubramani@, @rok987

**Comment by stephenplusplus** [comment]
Updated-- we've been good on Windows for a while. We're also making sure changes to the way our testing process works when we modularize won't affect Windows contributors (https://github.com/GoogleCloudPlatform/gcloud-node/pull/1431#issuecomment-233499000).
